[WebXR] Avoid re-allocation of metal textures for every frame

achan00 opened the pull request:
#### 32a1e991f51741f6d02c5025fd3773ffcd9302c6
<pre>
[WebXR] Avoid re-allocation of metal textures for every frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=274882">https://bugs.webkit.org/show_bug.cgi?id=274882</a>
<a href="https://rdar.apple.com/128010992">rdar://128010992</a>

Reviewed by Dan Glastonbury and Mike Wyrzykowski.

Before this code change, metal texture handles are passed in
the frame data for every frame render, and we&apos;d allocate a metal
texture from this handle (and later deallocate it) for each frame.

With this change, we allow textures to be reused. PlatformXRSystem
can send back a reusable index along with the texture handle, and
WebXROpaqueFramebuffer caches the allocated texture with that index.
When this same texture can be reused in a future frame, PlatformXRSystem
will send back that index in the frame data with a null texture handle
as a signal that the last texture stored at that index should be used.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
(WebCore::WebXRExternalRenderbuffer::operator bool const):
(WebCore::WebXRAttachmentSet::operator bool const):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
The cleanup of the display attachments is now handled in releaseAllDisplayAttachments().
(WebCore::WebXROpaqueFramebuffer::startFrame):
Call the new helper method bindCompositorTexturesForDisplay() that will
reuse a cached texture if possible.
(WebCore::WebXROpaqueFramebuffer::endFrame):
We no longer need to deallocate the texture after each frame render.
(WebCore::WebXROpaqueFramebuffer::releaseAllDisplayAttachments):
Release all the cached display attachments.
(WebCore::WebXROpaqueFramebuffer::blitShared):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachments const):
If texture handle is null, return the texture cached at the reusableTextureIndex.
(WebCore::WebXROpaqueFramebuffer::bindCompositorTexturesForDisplay):
First, ensure the m_displayAttachmentsSets vector is large enough to store
the set of display attachments used for this frame render. If there&apos;s no
layerData.textureData, which is the case when we run layout tests, create a
temporary renderbuffer to be used as the display color buffer. Otherwise,
check if we can reuse a previously created texture. If reuse is not possible,
create new textures with the handles passed in the frame data.
(WebCore::WebXROpaqueFramebuffer::releaseDisplayAttachmentsAtIndex):
Helper method to clean up the display attachments, refactored from old code.
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::didCompleteShutdown):
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
(WebCore::WebXRWebGLLayer::sessionEnded):
Clean up any cached display attachments.
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h:
* Source/WebCore/platform/xr/PlatformXR.h:
Introduce ExternalTextureData that packages up all the texture info
along with the reusable index.
Store LayerData with UniqueRef so we don&apos;t have to worry about its
size getting too large to be hashable.
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.cpp:
(PlatformXR::OpenXRDevice::requestFrame):
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
Don&apos;t pass any textureData from the fake device so a temporary
renderbuffer will be used for the color display buffer.
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::renderLoop):

Canonical link: <a href="https://commits.webkit.org/279523@main">https://commits.webkit.org/279523@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2cb050bbbe065f095b3ed1c1f9a58e9fa0b3f24d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33038 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6189 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56954 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4399 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55977 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4231 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43483 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2870 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46406 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24618 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3727 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2554 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3905 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58548 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28836 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3945 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50888 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30036 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46570 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50233 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11703 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30968 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29812 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg reviewed (approved):
:shipit:

mwyrzykowski commented on Source/WebKit/Shared/XR/PlatformXR.serialization.in:
```diff
@@ -181,7 +186,7 @@ header: <WebCore/PlatformXR.h>
     PlatformXR::FrameData::StageParameters stageParameters;
     Vector<PlatformXR::FrameData::View> views;
-    HashMap<PlatformXR::LayerHandle, PlatformXR::FrameData::LayerData> layers;
+    HashMap<PlatformXR::LayerHandle, UniqueRef<PlatformXR::FrameData::LayerData>> layers;
```
👍

cdumez commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
```diff
@@ -419,6 +404,79 @@ bool WebXROpaqueFramebuffer::setupFramebuffer(GraphicsContextGL& gl, const Platf
+
+    auto size = m_displayAttachmentsSets[index].size();
+    RefPtr gl = m_context.graphicsContextGL();
+    for (size_t i = 0; i < size; ++i) {
```
nit: You may want to use a modern range loop here:
```
for (auto& attachments : m_displayAttachmentsSets[index]) {
    if (gl)
        attachments.release(*gl);
    else
        attachments.leakObject();
}
```

cdumez commented on Source/WebCore/platform/xr/PlatformXR.h:
```diff
@@ -278,14 +278,20 @@ struct FrameData {
         uint64_t renderingFrameIndex { 0 };
-        ExternalTexture colorTexture = { MachSendRight(), false };
-        ExternalTexture depthStencilBuffer = { MachSendRight(), false };
+        std::optional<ExternalTextureData> textureData = { std::nullopt };
```
I doubt `= { std::nullopt }` is needed here.

achan00 committed:
[WebXR] Avoid re-allocation of metal textures for every frame
https://bugs.webkit.org/show_bug.cgi?id=274882
rdar://128010992

Reviewed by Dan Glastonbury and Mike Wyrzykowski.

Before this code change, metal texture handles are passed in
the frame data for every frame render, and we'd allocate a metal
texture from this handle (and later deallocate it) for each frame.

With this change, we allow textures to be reused. PlatformXRSystem
can send back a reusable index along with the texture handle, and
WebXROpaqueFramebuffer caches the allocated texture with that index.
When this same texture can be reused in a future frame, PlatformXRSystem
will send back that index in the frame data with a null texture handle
as a signal that the last texture stored at that index should be used.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
(WebCore::WebXRExternalRenderbuffer::operator bool const):
(WebCore::WebXRAttachmentSet::operator bool const):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
The cleanup of the display attachments is now handled in releaseAllDisplayAttachments().
(WebCore::WebXROpaqueFramebuffer::startFrame):
Call the new helper method bindCompositorTexturesForDisplay() that will
reuse a cached texture if possible.
(WebCore::WebXROpaqueFramebuffer::endFrame):
We no longer need to deallocate the texture after each frame render.
(WebCore::WebXROpaqueFramebuffer::releaseAllDisplayAttachments):
Release all the cached display attachments.
(WebCore::WebXROpaqueFramebuffer::blitShared):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachments const):
If texture handle is null, return the texture cached at the reusableTextureIndex.
(WebCore::WebXROpaqueFramebuffer::bindCompositorTexturesForDisplay):
First, ensure the m_displayAttachmentsSets vector is large enough to store
the set of display attachments used for this frame render. If there's no
layerData.textureData, which is the case when we run layout tests, create a
temporary renderbuffer to be used as the display color buffer. Otherwise,
check if we can reuse a previously created texture. If reuse is not possible,
create new textures with the handles passed in the frame data.
(WebCore::WebXROpaqueFramebuffer::releaseDisplayAttachmentsAtIndex):
Helper method to clean up the display attachments, refactored from old code.
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::didCompleteShutdown):
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
(WebCore::WebXRWebGLLayer::sessionEnded):
Clean up any cached display attachments.
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h:
* Source/WebCore/platform/xr/PlatformXR.h:
Introduce ExternalTextureData that packages up all the texture info
along with the reusable index.
Store LayerData with UniqueRef so we don't have to worry about its
size getting too large to be hashable.
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.cpp:
(PlatformXR::OpenXRDevice::requestFrame):
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
Don't pass any textureData from the fake device so a temporary
renderbuffer will be used for the color display buffer.
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::renderLoop):

Canonical link: https://commits.webkit.org/279523@main
